src/inventory/weapons/InstantWeapon.cpp

```diff
@@ -1,43 +1,75 @@
-/*
-    Created  by Deric M       01/03/17
+/**
+    InstantWeapon.cpp
+
+    DISCRIPTION:
+        InstantWewapon Weapons construct a line from the weapons mussle to its
+        theoretical range limit, and then get all the intersecting targets.
+        Tergets are sorted in a priority queue by distance from player, and then
+        they are damaged in order untill something invulnrable is hit, or
+        penertation runs out.
+
+    AUTHOR: Deric Mccadden 01/03/17
+
 */
+#include <queue>
+#include <stdio.h>
+#include <iostream>
+
 #include "InstantWeapon.h"
 #include "../../collision/HitBox.h"
 #include "../../game/GameManager.h"
 #include "../../collision/CollisionHandler.h"
 #include "../../audio/AudioManager.h"
-#include <queue>
-#include <stdio.h>
-#include <iostream>
 #include "../../log/log.h"
+#include "Target.h"
 #include "../../sprites/VisualEffect.h"
 
-InstantWeapon::InstantWeapon(std::string type, int range, int damage,
-        int clip, int clipMax, int ammo, int AOE, int reloadSpeed, int fireRate, bool isReadyToFire)
-        : Weapon(type, range, damage, clip, clipMax, ammo, AOE, reloadSpeed, fireRate, isReadyToFire) {
+using std::string;
 
-}
+InstantWeapon::InstantWeapon(string type, string fireSound, string hitSound, string reloadSound, string emptySound,
+        int range, int damage, int AOE, int penetration, int clip, int clipMax, int ammo, int reloadDelay, int fireDelay)
+: Weapon(type, fireSound, hitSound, reloadSound, emptySound, range, damage, AOE, penetration, clip, clipMax, ammo,
+        reloadDelay, fireDelay) {
 
+}
 
 
 // DericM, 01/03/17
-void InstantWeapon::fire(Marine &marine){
- //check ammo
+bool InstantWeapon::fire(Marine &marine) {
 
+    if(!Weapon::fire(marine)) {
+        return false;
+    }
     logv("InstantWeapon::fire()\n");
 
-    CollisionHandler &collisionHandler = GameManager::instance()->getCollisionHandler();
+    std::priority_queue<Target> targets;
+    GameManager* gameManager = GameManager::instance();
+    auto& collisionHandler = gameManager->getCollisionHandler();
 
-    if(!reduceAmmo(1)){
-        return;
-    }
+    collisionHandler.detectLineCollision(targets, marine, range);
+
+    for(int i = 0; i < penetration; i++) {
+        if(targets.empty()) {
```

Comment:
Targets check should be performed before the loop
Otherwise if it is not empty, then it will just keep checking for no reason